### PR TITLE
[sw,tests] Chip level test `chip_sw_aon_timer_wdog_bite_reset`

### PR DESCRIPTION
**Pull request by doreis-lowrisc**
# This PR add two tests listed in the chip-level testplan:
## chip_sw_aon_timer_wdog_bite_reset:
 - Read the reset cause register in rstmgr to confirm that the SW is in the POR reset phase.
 - Program the AON timer wdog to 'bark' after some time.
 - Let the bark escalate to bite, which should result in a reset request.
 - After reset, read the reset cause register in rstmgr to confirm that the SW is now in the wdog reset phase.

## chip_sw_aon_timer_sleep_wdog_bite_reset:
- Repeat the steps in chip_aon_timer_wdog_bite_reset test, but with following changes:
- Program the pwrmgr to go to deep sleep state (clocks off, power off).
- Issue a WFI after programming the wdog, so that the reset request due to bite occurs during deep sleep state.
- After reset, read the reset cause register in rstmgr to confirm that the SW is now in the wdog reset phase.

### CODE REVIEW

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (moidx) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:42]: A recommendation to make the checks more explicit: 

1. Set `bark_cycles` and `bite_cycles`
2. Call `usleep(n)` making sure `n > bark_time_us`.
3. Check `kDifAonTimerIrqWdogTimerBark` using `dif_aon_timer_irq_is_pending`

This way the test will ensure that the bark signal was triggered.

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:21]: Nit: D**e**finitions? Also, can you open an issue about this and tag it here? Or just manually plonk them in `dif_rstmgr.h` for now?

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:92]: At the moment, we don't check that this isn't e.g. `RSTMGR_REASON_POWERON | RSTMGR_REASON_POWER_UNSTABLE`. Should this be something more like `rst_info == RSTMGR_REASON_WATCHDOG || rst_info == RSTMGR_REASON_POWERON` to make sure we've got exactly one bit set, which is one of the bits we're expecting?

(And, if so, we can change the `if` tests below to `==`)

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:104]: Missing newline?

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:43]: Should the intermediate product be stored in a `uint64_t`? I think bad things might happen if `bark_time_us` is big.

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:57]: Nit: This should probably be "... check **that** the ..." (since only one answer is ok!)

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:65]: Maybe throw in an `assert (bark_time_us < bite_time_us)` to check that nothing overflowed?

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:67]: Nit: This should probably be "... waiting **for** ..."

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (rswarbrick) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: I'd suggest you make that change here and wait for review. Otherwise you can end up in a situation where no-one does anything because they don't realise you're waiting for them to reply.

Comment (doreis-lowrisc) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: Done!

**Review by rswarbrick — APPROVED** [review]
Looks good to me, but we probably need a SW review before merging.

**Comment by silvestrst** [comment]
> Looks good to me, but we probably need a SW review before merging.

I will give it a quick passthrough now.

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (silvestrst) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:79]: Here and elsewhere, is there a reason why you decided not to use `dif_rstmgr_reset_info_t` variants?

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (silvestrst) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:85]: Could be a `#define`

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (silvestrst) [inline sw/device/tests/aon_timer_wdog_bite_reset_test.c:44]: Nit: s/interrupt fired/interrupt requested/

**Review by silvestrst — COMMENTED** [review]
@doreis-lowrisc thanks!

File: sw/device/lib/dif/dif_rstmgr.h
Comment (doreis-lowrisc) [reply sw/device/lib/dif/dif_rstmgr.h:54]: As described in the documentation the `dif_rstmgr_reset_info_t` doesn't define the different Hardware reasons. I think that the best solution will be to split the `kDifRstmgrResetInfoHwReq`.

Comment (silvestrst) [reply sw/device/lib/dif/dif_rstmgr.h]: I think that defining these defines here is not the right thing to do, otherwise they probably would have been named register bits.

I think maybe the better place would be `sw/device/lib/testing/rstmgr_testutils.<c,h>`.

@mundaym what do you think?

File: sw/device/tests/aon_timer_wdog_bite_reset_test.c
Comment (doreis-lowrisc) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: The difs which uses these pointers doesn't define them as const, if I do so I would need to cast them. So I think I'll leave it this way.

Comment (silvestrst) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: I think the only DIF functions that don't have `const` are `dif_<peripheral>_init`, if I am not mistaken, the rest will never change the contents of `dif_<peripheral>_t`, and hence should always be `const` pointers.

The only places that I can see don't use `const` are functions in the testutils that you use in this code.

It could be a nice clean-up to add `const` to them (fairly minimal) change.

This is really a minor thing, but just to be consistent it could be a nice little tidy-up.

Comment (doreis-lowrisc) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: You're right, is `aon_timer_testutils_wakeup_config` functions that is not defining it as const.

Comment (silvestrst) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: It's up to you, if you decide to go for it, I think it makes sense in a separate clean-up PR.

Comment (doreis-lowrisc) [reply sw/device/tests/aon_timer_wdog_bite_reset_test.c]: Done in a different commit.

File: sw/device/lib/dif/dif_rstmgr.h
Comment (doreis-lowrisc) [reply sw/device/lib/dif/dif_rstmgr.h]: You can have a look how this was approached on mask_rom
https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/rstmgr.h;l=37

Comment (silvestrst) [reply sw/device/lib/dif/dif_rstmgr.h]: I guess there are two things:
1) Whether we should define new variants
2) Eliminating redundancy

1) I guess it's fine to mimic the Mask ROM implementation
2) In my opinion, should either keep the defines or the existing `dif_rstmgr_reset_info_t`. Otherwise there are essentially two copies of the same thing - one more complete, the other one less complete. As `dif_rstmgr_reset_info_t` did exist before the defines, it seems reasonable to extend it. Alternatively, the DIF could match the HW exactly, and then have additional variants in `testutils`.

**Comment by doreis-lowrisc** [comment]
@rswarbrick Sorry about that, but I did a slight change to cover also the test `chip_sw_aon_timer_sleep_wdog_bite_reset`.

**Review by rswarbrick — APPROVED** [review]
I think I'd probably squash the second two commits, but don't have strong feelings. As before, let's get the approval of a SW person before merge.

**Review by moidx — APPROVED** [review]

**Pull request merged**
